File: go/mac/hmac_key_manager.go
// Copyright 2018 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
////////////////////////////////////////////////////////////////////////////////

package mac

import (
	"errors"
	"fmt"

	"google.golang.org/protobuf/proto"
	"github.com/google/tink/go/keyset"
	"github.com/google/tink/go/mac/subtle"
	"github.com/google/tink/go/subtle/random"
	commonpb "github.com/google/tink/go/proto/common_go_proto"
	hmacpb "github.com/google/tink/go/proto/hmac_go_proto"
	tinkpb "github.com/google/tink/go/proto/tink_go_proto"
)

const (
	hmacKeyVersion = 0
	hmacTypeURL    = "type.googleapis.com/google.crypto.tink.HmacKey"
)

var errInvalidHMACKey = errors.New("hmac_key_manager: invalid key")
var errInvalidHMACKeyFormat = errors.New("hmac_key_manager: invalid key format")

// hmacKeyManager generates new HMAC keys and produces new instances of HMAC.
type hmacKeyManager struct{}

// Primitive constructs a HMAC instance for the given serialized HMACKey.
func (km *hmacKeyManager) Primitive(serializedKey []byte) (interface{}, error) {
	if len(serializedKey) == 0 {
		return nil, errInvalidHMACKey
	}
	key := new(hmacpb.HmacKey)
	if err := proto.Unmarshal(serializedKey, key); err != nil {
		return nil, errInvalidHMACKey
	}
	if err := km.validateKey(key); err != nil {
		return nil, err
	}
	hash := commonpb.HashType_name[int32(key.Params.Hash)]
	hmac, err := subtle.NewHMAC(hash, key.KeyValue, key.Params.TagSize)
	if err != nil {
		return nil, err
	}
	return hmac, nil
}

// NewKey generates a new HMACKey according to specification in the given HMACKeyFormat.
func (km *hmacKeyManager) NewKey(serializedKeyFormat []byte) (proto.Message, error) {
	if len(serializedKeyFormat) == 0 {
		return nil, errInvalidHMACKeyFormat
	}
	keyFormat := new(hmacpb.HmacKeyFormat)
	if err := proto.Unmarshal(serializedKeyFormat, keyFormat); err != nil {
		return nil, errInvalidHMACKeyFormat
	}
	if err := km.validateKeyFormat(keyFormat); err != nil {
		return nil, fmt.Errorf("hmac_key_manager: invalid key format: %s", err)
	}
	keyValue := random.GetRandomBytes(keyFormat.KeySize)
	return &hmacpb.HmacKey{
		Version:  hmacKeyVersion,
		Params:   keyFormat.Params,
		KeyValue: keyValue,
	}, nil
}

// NewKeyData generates a new KeyData according to specification in the given
// serialized HMACKeyFormat. This should be used solely by the key management API.
func (km *hmacKeyManager) NewKeyData(serializedKeyFormat []byte) (*tinkpb.KeyData, error) {
	key, err := km.NewKey(serializedKeyFormat)
	if err != nil {
		return nil, err
	}
	serializedKey, err := proto.Marshal(key)
	if err != nil {
		return nil, errInvalidHMACKeyFormat
	}

	return &tinkpb.KeyData{
		TypeUrl:         hmacTypeURL,
		Value:           serializedKey,
		KeyMaterialType: tinkpb.KeyData_SYMMETRIC,
	}, nil
}

// DoesSupport checks whether this KeyManager supports the given key type.
func (km *hmacKeyManager) DoesSupport(typeURL string) bool {
	return typeURL == hmacTypeURL
}

// TypeURL returns the type URL of keys managed by this KeyManager.
func (km *hmacKeyManager) TypeURL() string {
	return hmacTypeURL
}

// validateKey validates the given HMACKey. It only validates the version of the
// key because other parameters will be validated in primitive construction.
func (km *hmacKeyManager) validateKey(key *hmacpb.HmacKey) error {
	err := keyset.ValidateKeyVersion(key.Version, hmacKeyVersion)
	if err != nil {
		return fmt.Errorf("hmac_key_manager: invalid version: %s", err)
	}
	keySize := uint32(len(key.KeyValue))
	hash := commonpb.HashType_name[int32(key.Params.Hash)]
	return subtle.ValidateHMACParams(hash, keySize, key.Params.TagSize)
}

// validateKeyFormat validates the given HMACKeyFormat
func (km *hmacKeyManager) validateKeyFormat(format *hmacpb.HmacKeyFormat) error {
	if format.Params == nil {
		return fmt.Errorf("null HMAC params")
	}
	hash := commonpb.HashType_name[int32(format.Params.Hash)]
	return subtle.ValidateHMACParams(hash, format.KeySize, format.Params.TagSize)
}

File: javascript/mac/hmac_key_manager.ts
/**
 * @license
 * Copyright 2023 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */

import {SecurityException} from '../exception/security_exception';
import * as keyManager from '../internal/key_manager';
import {PbHashType, PbHmacKey, PbHmacKeyFormat, PbHmacParams, PbKeyData, PbMessage} from '../internal/proto';
import {bytesAsU8, bytesLength} from '../internal/proto_shims';
import * as registry from '../internal/registry';
import {Constructor} from '../internal/util';
import * as hmac from '../subtle/hmac';
import * as random from '../subtle/random';
import * as validators from '../subtle/validators';

import {Mac} from './internal/mac';

const VERSION = 0;
const SUPPORTED_PRIMITIVE = Mac;
const MIN_KEY_SIZE = 16;
const MIN_TAG_SIZE = 10;
const MAX_TAG_SIZE = new Map([
  [PbHashType.SHA1, 20], [PbHashType.SHA256, 32], [PbHashType.SHA384, 48],
  [PbHashType.SHA512, 64]
]);

/**
 * A container for methods that generate new HMAC keys.
 * @final
 */
class HmacKeyFactory implements keyManager.KeyFactory {
  /**
   * Generates a new random HMAC key according to 'keyFormat'.
   *
   * @param keyFormat is either a KeyFormat
   *     proto or a serialized KeyFormat proto
   * @return the new generated HMAC key
   */
  newKey(keyFormat: PbMessage|Uint8Array) : PbHmacKey {
    const keyFormatProto = getKeyFormatProto(keyFormat);
    const {hmacParams, hmacKeySize} = this.validateKeyFormat(keyFormatProto);

    return new PbHmacKey()
        .setVersion(VERSION)
        .setParams(hmacParams)
        .setKeyValue(random.randBytes(hmacKeySize));
  }

  /**
   * Generates a new random HMAC key based on the `serializedKeyFormat` and
   * returns it as a KeyData proto.
   */
  newKeyData(serializedKeyFormat: Uint8Array): PbKeyData {
    const key: PbHmacKey = this.newKey(serializedKeyFormat);
    return new PbKeyData()
        .setTypeUrl(HmacKeyManager.KEY_TYPE)
        .setValue(key.serializeBinary())
        .setKeyMaterialType(PbKeyData.KeyMaterialType.SYMMETRIC);
  }

  validateKeyFormat(keyFormat: PbHmacKeyFormat):
      {hmacParams: PbHmacParams, hmacKeySize: number} {
    validators.validateVersion(keyFormat.getVersion(), VERSION);
    if (!keyFormat.getKeySize()) {
      throw new SecurityException('Invalid HMAC key format: key size not set');
    }
    if (keyFormat.getKeySize() < MIN_KEY_SIZE) {
      throw new SecurityException(
          `Key too short, must be at least ${MIN_KEY_SIZE} bytes.`);
    }
    const params: PbHmacParams|undefined = keyFormat.getParams();
    if (!params) {
      throw new SecurityException('Invalid HMAC key format: params not set');
    }
    const tagSize = params.getTagSize();
    if (!tagSize) {
      throw new SecurityException('Invalid HMAC params: tag size not set');
    }
    if (tagSize < MIN_TAG_SIZE) {
      throw new SecurityException(
          'Invalid HMAC params: tag size ' + String(tagSize) +
          ' is too small.');
    }
    const hashType = params.getHash();
    if (!MAX_TAG_SIZE.has(hashType)) {
      throw new SecurityException('Invalid HMAC params: unknown hash type');
    }
    if (tagSize > MAX_TAG_SIZE.get(hashType)!) {
      throw new SecurityException(
          'Invalid HMAC params: tag size ' + String(tagSize) +
          ' is too large.');
    }

    return {
      hmacParams: params,
      hmacKeySize: keyFormat.getKeySize(),
    };
  }
}

  /**
   * The input keyFormat is either deserialized (in case that the input is
   * Uint8Array) or checked to be an HmacKeyFormat-proto (otherwise).
   *
   */
function getKeyFormatProto(keyFormat: PbMessage|Uint8Array): PbHmacKeyFormat {
  if (keyFormat instanceof Uint8Array) {
    return deserializeKeyFormat(keyFormat);
    // Used because of the internal vs. external protobuf API discrepancies.
    // tslint:disable-next-line:jspb-casts
  } else if (keyFormat instanceof PbHmacKeyFormat) {
    return keyFormat;
  } else {
    throw new SecurityException('Expected HmacKeyFormat-proto');
  }
}

function deserializeKeyFormat(keyFormat: Uint8Array): PbHmacKeyFormat {
  let keyFormatProto: PbHmacKeyFormat;
  try {
    keyFormatProto = PbHmacKeyFormat.deserializeBinary(keyFormat);
  } catch (e) {
    throw new SecurityException(
        'Could not parse the input as a serialized proto of ' +
        HmacKeyManager.KEY_TYPE + ' key format.');
  }
  if (!keyFormatProto.getKeySize() || !keyFormatProto.getParams()) {
    throw new SecurityException(
        'Could not parse the input as a serialized proto of ' +
        HmacKeyManager.KEY_TYPE + ' key format.');
  }
  return keyFormatProto;
}


/**
 * Key manager that generates new {@link HmacKey} keys and produces new
 * instances of {@link Mac} primitives.
 *
 * @final
 */
export class HmacKeyManager implements keyManager.KeyManager<Mac> {
  static KEY_TYPE = 'type.googleapis.com/google.crypto.tink.HmacKey';
  private readonly keyFactory = new HmacKeyFactory();

  async getPrimitive(primitiveType: Constructor<Mac>, key: PbKeyData|PbMessage):
      Promise<Mac> {
    if (primitiveType !== this.getPrimitiveType()) {
      throw new SecurityException(
          'Requested primitive type which is not supported by this key manager.');
    }
    const keyProto = getKeyProto(key);
    const {hashType, keyValue, tagSize} = this.validateKey(keyProto);
    return await hmac.fromRawKey(hashType, keyValue, tagSize);
  }

  doesSupport(keyType: string) {
    return keyType === this.getKeyType();
  }

  getKeyType() {
    return HmacKeyManager.KEY_TYPE;
  }

  getPrimitiveType() {
    return SUPPORTED_PRIMITIVE;
  }

  getVersion() {
    return VERSION;
  }

  getKeyFactory() {
    return this.keyFactory;
  }

  validateKey(key: PbHmacKey):
      {hashType: string, keyValue: Uint8Array, tagSize: number} {
    validators.validateVersion(key.getVersion(), VERSION);
    const keyFormat = (new PbHmacKeyFormat())
                          .setParams(key.getParams())
                          .setKeySize(bytesLength(key.getKeyValue()));
    const {hmacParams} = this.keyFactory.validateKeyFormat(keyFormat);
    let hashType: string;
    switch (hmacParams.getHash()) {
      case PbHashType.SHA1:
        hashType = 'SHA-1';
        break;
      case PbHashType.SHA256:
        hashType = 'SHA-256';
        break;
      case PbHashType.SHA384:
        hashType = 'SHA-384';
        break;
      case PbHashType.SHA512:
        hashType = 'SHA-512';
        break;
      default:
        throw new SecurityException('Unknown hash type');
    }
    return {
      hashType,
      keyValue: bytesAsU8(key.getKeyValue()),
      tagSize: hmacParams.getTagSize()
    };
  }

  static register() {
    registry.registerKeyManager(new HmacKeyManager());
  }
}

  /**
   * The input key is either deserialized (in case that the input is
   * KeyData-proto) or checked to be an HmacKey-proto (otherwise).
   *
   */
function getKeyProto(keyMaterial: PbMessage|PbKeyData): PbHmacKey {
  // tslint:disable:jspb-casts
  if (keyMaterial instanceof PbKeyData) {
    return getKeyProtoFromKeyData(keyMaterial);
  } else if (keyMaterial instanceof PbHmacKey) {
    return keyMaterial;
  } else {
    throw new SecurityException(
        'Key type is not supported. This key manager supports ' +
        HmacKeyManager.KEY_TYPE + '.');
  }  // tslint:enable:jspb-casts
}

  /**
   * It validates the key type and returns a deserialized HmacKey-proto.
   *
   */
function getKeyProtoFromKeyData(keyData: PbKeyData): PbHmacKey {
  if (keyData.getTypeUrl() !== HmacKeyManager.KEY_TYPE) {
    throw new SecurityException(
        'Key type ' + keyData.getTypeUrl() +
        ' is not supported. This key manager supports ' +
        HmacKeyManager.KEY_TYPE + '.');
  }
  let deserializedKey: PbHmacKey;
  try {
    deserializedKey = PbHmacKey.deserializeBinary(keyData.getValue());
  } catch (e) {
    throw new SecurityException(
        'Could not parse the input as a ' +
        'serialized proto of ' + HmacKeyManager.KEY_TYPE + ' key.');
  }
  return deserializedKey;
}
